subsys/net/lib/sockets/sockets.c

```diff
@@ -1764,6 +1764,45 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 			}
 
 			break;
+
+		case SO_BINDTODEVICE: {
+			struct net_if *iface;
+			const struct device *dev;
+			struct ifreq *ifreq = (struct ifreq *)optval;
+
+			/* optlen equal to 0 or empty interface name should
+			 * remove the binding.
+			 */
+			if ((optlen == 0) || (ifreq != NULL &&
+					      strlen(ifreq->ifr_name) == 0)) {
+				ctx->flags &= ~NET_CONTEXT_BOUND_TO_IFACE;
+				return 0;
+			}
+
+			if ((ifreq == NULL) || (optlen != sizeof(*ifreq))) {
+				errno = EINVAL;
+				return -1;
+			}
+
+			dev = device_get_binding(ifreq->ifr_name);
+			if (dev == NULL) {
+				errno = ENODEV;
+				return -1;
+
+			}
+
+			iface = net_if_lookup_by_dev(dev);
```

Comment:
This works if one dev has only one interface, but for VLAN's there can be multiple network interfaces tied to a one specific device. So in that case, the end result might be unintentional. On the other hand, the interface selection is most probably not done using this socket option if VLAN is enabled. This can be investigated more later if this becomes a real issue.

Reply:
Hmm, I have not thought about the virtual interfaces. Do you have any idea how could we address this, shall we introduce a `name` field to the `net_if` structure and `net_if_lookup_by_name()`-like functionality? This sounds like a heavy change, if it could be postponed until the actual problem arises with virtual interfaces it would be good.

Reply:
> if it could be postponed until the actual problem arises with virtual interfaces it would be good.

Yes, I agree. Lets tackle this issue when we get there. I think this PR is a good solution atm.

Reply:
BTW, we could use the interface index as a name, so there should not be need to allocate extra spece for interface name.